src/test/libcephfs/reclaim.cc

```diff
@@ -0,0 +1,79 @@
+// -*- mode:C++; tab-width:8; c-basic-offset:2; indent-tabs-mode:t -*-
+// vim: ts=8 sw=2 smarttab
+/*
+ * Tests for Ceph delegation handling
+ *
+ * (c) 2017, Jeff Layton <jlayton@redhat.com>
+ */
+
+#include "gtest/gtest.h"
+#include "include/cephfs/libcephfs.h"
+#include "include/stat.h"
+#include <errno.h>
+#include <fcntl.h>
+#include <unistd.h>
+#include <sys/types.h>
+#include <sys/stat.h>
+#include <dirent.h>
+#include <sys/xattr.h>
+#include <sys/uio.h>
+#include <libgen.h>
+
+#ifdef __linux__
+#include <limits.h>
+#endif
+
+#include <map>
+#include <vector>
+#include <thread>
+#include <atomic>
+
+#define	CEPHFS_RECLAIM_TIMEOUT		60
+
+TEST(LibCephFS, ReclaimReset) {
+  pid_t		pid;
+  char		uuid[256];
+
+  sprintf(uuid, "simplereclaim:%x", getpid());
+
+  pid = fork();
+  ASSERT_GE(pid, 0);
+  if (pid == 0) {
+    int len;
+    char child[PATH_MAX];
+    char *env = getenv("EXECUTABLE_NAME");
+
+    len = snprintf(child, PATH_MAX, "%s/ceph_test_libcephfs_dying_client",
+		   dirname(env));
```

Comment:
It has been a while since I wrote the test, but I think I was concerned about things happening in the "background" in the child's mount. Just because we're not doing any client-side activity on that mount, things could still be happening on it (cap revokes/releases, etc). Also, there are things associated with a libcephfs mount that are truly global in scope. Doing this with separate processes ensures that none of that stuff is shared.

That said, you have a good point. We probably should also test the case where the first client is still alive and a second client issues reclaim or kills that session. That does need to be handled in some sane fashion (probably by shutting down the first mount). I think we probably ought to roll a separate test for that?

Reply:
Yes, a separate test sounds appropriate. You're right that trying to reclaim another mount in the same process might lead to problems (which aren't really worth testing/fixing). How about have the test executable you wrote handle both situations?

We should expect that reclaiming a session causes the previous instance to be (a) blacklisted and (b) reconnects to the original session are denied.